### PR TITLE
Get gift card data

### PR DESCRIPTION
**Pull request by atorresveiga**
⚠️ This PR depends on this tiny [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2696)

Closes: #8398

### Why
We are adding gift card read-only support to the Woo App

### Description
This PR adds the necessary changes _to fetch the gift card information from the API_ . Actually, it doesn't bring the whole gift card information from the API, but relies only on the summary gift card information the orders request contains. It also adds a minor change in the decision-making logic of whether fetch hosted woo extension information or not. Now the extension should appear installed and active in the local database to fetch the data.

### Testing instructions
Prerequisites:
* Install the gift card extension in your store and place some orders where you apply/redeem some gift cards

TC1 
1. Open the orders tab
2. Select an order with a gift card applied/redeemed
3. Check that the gift card information appears on the order details (The Gift Card section and the Payment Section)

TC2
1. Open the orders tab
2. Select an order where no gift card was applied/redeemed
3.  Check that NO gift card information appears

TC3
1. Deactivate the Gift Card extension
2. Close/Open the app (to refresh the list of plugins)
3. Open the orders tab
4. Select an order with a gift card applied/redeemed
5. Check that NO gift card information appears (because the plugin is deactivated)

### Images/gif

https://user-images.githubusercontent.com/18119390/229257516-9ba76efa-8d31-4f9f-86a3-c79d66a85437.mp4

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/network/giftcard/GiftCardRestClient.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            params = params
+        ).toWooPayload { giftCardSummaryResponse ->
+            giftCardSummaryResponse.gift_cards ?: emptyList()
+        }
+    }
+
+    @Suppress("PropertyName", "VariableNaming")
+    class GiftCardSummaryResponse : Response {
+        val gift_cards: List<GiftCardSummaryDto>? = null
+    }
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/network/giftcard/GiftCardRestClient.kt]: It's not a blocking suggestion. 

You could use `@SerializedName` annotation + I think this class doesn't have to inherit from `Response` and could be a `data class`.

```suggestion
    data class GiftCardSummaryResponse (
        @SerializedName("gift_cards") val giftCards: List<GiftCardSummaryDto>? = null
    )
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt [hunk 2]
```diff
@@ -693,6 +680,18 @@ class OrderDetailViewModel @Inject constructor(
         }
     }
 
+    private suspend fun fetchGiftCardsAsync() = async {
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:681]: ⚙️ WDYT about adding support for `OrderDetailsTransactionLauncher` in both `fetchSubscriptionsAsync` and `fetchGiftCardsAsync` methods? As they are needed for rendering the screen for some orders, they should be included in the transaction I believe.

**Review by wzieba — APPROVED** [review]
Looks good! No blocking comments.

Leaving this unmerged as I'm not 100% sure how would you prefer to approach updating the FluxC version.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt (hunk 2 above)
Comment (atorresveiga) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:681]: Good point! Fixed [here](https://github.com/woocommerce/woocommerce-android/pull/8712/commits/b65050b9b59eabdb38bbe4567609f9b06e28d606)

**Comment by wzieba** [comment]
Thanks for addressing the transaction improvements 👍

I just realized that you've added fetching subscriptions and gift card data into a separate `awaitAll`. Could we consider making all calls parallel so the overall time of loading of an order will be shorter?

image

**Comment by wzieba** [comment]
More historical context on the above suggestion ☝️ is available here (internal): pe5pgL-Gh-p2

**Comment by atorresveiga** [comment]
> Could we consider making all calls parallel so the overall time of loading of an order will be shorter?

@wzieba Considering that some extensions need values from the order metadata, I decided to use a different parallel call to fetch all of the hosted Woo extensions once we have fetched the core order values. This would help to ensure that we have refreshed the cached order when analyzing the metadata. Please note that the Order will be displayed as soon as we fetch the core data. Does that make sense?

**Comment by wzieba** [comment]
Frankly, I'm not sure if I follow but at the same time I don't want to block you with this PR (I'm going AFK for the rest of the week).

Please feel free to merge this PR! We can always discuss it sometime later.

---

I'll leave the thought for future discussion: as some extensions don't seem to need order's metadata (e.g. discussed here GiftCard needs only order id), I'd vote to parallelize all _possible_ calls. This way user will receive details of the Order ASAP.

E.g. if an order has some data only from GiftCard, there's no need to wait for e.g. Subscriptions (assuming that Subscriptions need metadata).

**Comment by atorresveiga** [comment]
>  I'd vote to parallelize all possible calls. This way user will receive details of the Order ASAP.

I will follow this approach above ☝️. Regarding gift cards fixed [here](https://github.com/woocommerce/woocommerce-android/pull/8712/commits/dca68ca151479364d90a71df34209116790578fb). Subscriptions will need that the order is synced first, though.

**Pull request merged**